beacon-chain/state/state-native/getters_validator.go

```diff
@@ -133,43 +157,84 @@ func (b *BeaconState) ValidatorIndexByPubkey(key [fieldparams.BLSPubkeyLength]by
 // PubkeyAtIndex returns the pubkey at the given
 // validator index.
 func (b *BeaconState) PubkeyAtIndex(idx primitives.ValidatorIndex) [fieldparams.BLSPubkeyLength]byte {
-	if uint64(idx) >= uint64(len(b.validators)) {
-		return [fieldparams.BLSPubkeyLength]byte{}
-	}
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
-	if b.validators[idx] == nil {
+	var v *ethpb.Validator
+	if features.Get().EnableExperimentalState {
+		var err error
+		v, err = b.validatorsMultiValue.At(b, uint64(idx))
+		if err != nil {
+			return [fieldparams.BLSPubkeyLength]byte{}
+		}
+	} else {
+		if uint64(idx) >= uint64(len(b.validators)) {
+			return [fieldparams.BLSPubkeyLength]byte{}
+		}
+		v = b.validators[idx]
+	}
+
+	if v == nil {
 		return [fieldparams.BLSPubkeyLength]byte{}
 	}
-	return bytesutil.ToBytes48(b.validators[idx].PublicKey)
+	return bytesutil.ToBytes48(v.PublicKey)
 }
 
 // NumValidators returns the size of the validator registry.
 func (b *BeaconState) NumValidators() int {
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
-	return len(b.validators)
+	return b.validatorsLen()
 }
 
 // ReadFromEveryValidator reads values from every validator and applies it to the provided function.
 //
 // WARNING: This method is potentially unsafe, as it exposes the actual validator registry.
 func (b *BeaconState) ReadFromEveryValidator(f func(idx int, val state.ReadOnlyValidator) error) error {
+	var validators []*ethpb.Validator
```

Comment:
is it possible to just make this code block a separate function ?

Reply:
What exactly do you want to extract? All of it?